### PR TITLE
Fix redeclaration related eval corner cases

### PR DESCRIPTION
**Pull request by szilagyiadam**
- Fix cases where a function parameter binding is redeclared inside a parameter initializer eval
- Fix cases where a let- or function arguments binding is redeclared inside a function block because
  there is a declaration in a function parameter initializer eval
- Also remove the ECMA_PARSE_CHAIN_INDEX_SHIFT macro, added a debugger_eval_chain_index named field to the jerry context instead

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/parser/js/js-scanner-util.c [hunk 1]
```diff
@@ -13,6 +13,8 @@
  * limitations under the License.
  */
 
+#include <string.h>
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-util.c:16]: is this really necessary?

File: jerry-core/parser/js/js-scanner-util.c [hunk 2]
```diff
@@ -2747,6 +2848,7 @@ scanner_create_variables (parser_context_t *context_p, /**< context */
           {
             case SCANNER_STREAM_TYPE_LET:
             {
+              //---------------
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-util.c:2851]: There are several comments here, please remove them.

**Review by rerobika — CHANGES_REQUESTED** [review]

File: jerry-core/parser/js/js-scanner-util.c [hunk 3]
```diff
@@ -400,8 +400,109 @@ scanner_seek (parser_context_t *context_p) /**< context */
...
+
+    uint8_t *destination_p = (uint8_t *) parser_malloc (context_p, lit_loc_p->length);
+
+    lexer_convert_ident_to_cesu8 (destination_p, lit_loc_p->char_p, lit_loc_p->length);
+
+    ecma_string_t *search_key_p = ecma_new_ecma_string_from_utf8 (destination_p, lit_loc_p->length);
+
+    scanner_free (destination_p, lit_loc_p->length);
+
+    if (ecma_compare_ecma_strings (arg_string, search_key_p))
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-util.c:494]: save to local, free `search_key_p` only once

File: jerry-core/parser/js/js-scanner-util.c [hunk 4]
```diff
@@ -400,8 +400,109 @@ scanner_seek (parser_context_t *context_p) /**< context */
...
+  if (scanner_literal_is_arguments (lit_loc_p))
+  {
+    return true;
+  }
+
+  uint16_t register_end, encoding_limit, encoding_delta;
+  ecma_value_t *literal_p;
+  ecma_value_t *literal_start_p;
+
+  const ecma_compiled_code_t *bytecode_header_p = vm_context_p->shared_p->bytecode_header_p;
```
Comment (rerobika) [inline jerry-core/parser/js/js-scanner-util.c:436]: Query `JERRY_CONTEXT (vm_top_context_p)` only here

**Review by rerobika — CHANGES_REQUESTED** [review]

**Comment by szilagyiadam** [comment]
Fixes #4896

**Review by rerobika — APPROVED** [review]

**Review by ossy-szeged — APPROVED** [review]

**Pull request merged**